src/CachePeers.h

```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright (C) 1996-2023 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID_CACHEPEERS_H
+#define SQUID_CACHEPEERS_H
+
+#include "base/forward.h"
+#include "CachePeer.h"
+#include "mem/PoolingAllocator.h"
+
+#include <vector>
+
+/// Weak pointers to zero or more Config.peers.
+/// Users must specify the selection algorithm and the order of entries.
+using SelectedCachePeers = std::vector< CbcPointer<CachePeer>, PoolingAllocator< CbcPointer<CachePeer> > >;
```

Comment:
The name for this type is bad "selected" is a use-specific detail that should be part of a local variable name, not a global type.

Reply:
> "selected" is a use-specific detail that should be part of a local variable name, not a global type.

When several users share the same detail, expressing that shared concern via a type name is often a good idea. In this particular case, it is a good idea to emphasize that this collection of CachePeer object is not a collection of _all_ configured cache_peers. This PR fixes at least one bug that is directly related to the difference between "all peers" and "selected peers". Moreover, for cbdata-protected types, the difference between "selected" and "all" cache_peers affects stored pointer type/validity and essentially _requires_ different container types (hence, different names).

This type name reflects those ideas, even though it cannot and does not not define how the selection is performed by each type user. Needless to say, a specific type user (that wants to reuse this type in several places) can create a user-specific alias that will identify the selection process by name. Doing so without duplicating code requires this type alias to exist.

Reply:
> This is a type for storing a non-singular amount of CachePeer - thus "CachePeers" is more correct.

If we had just one kind of CachePeer object collection, that conclusion would be correct. In reality, there are several different collections of CachePeer objects, including these three:

* `CachePeers` is the (future name) of the type representing `Config.peers`, `Config.npeers`, and associated functionality -- the object that owns all configured peers. Besides CachePeer ownership, it encapsulates configuration-related and peer polling functionality that other containers do not care about. This type does not exist as a named C++ type in the official code or this PR, but the underlying code does exist. It just lacks proper encapsulation. We are working on that, but we need to get this and several other "minimal" helper PRs out of the way first.

* `SelectedCachePeers` may not contain all configured peers and, today, it does not own any of them (because it is a collection of _weak_ cbdata pointers). Like CachePeers, it is suitable for long-term storage of peers, but without the configuration-related baggage that the CachePeers type in the first bullet has to carry.

* `RawCachePeers`, like SelectedCachePeers, may not contain all configured peers and does not own any of them (because it is a collection of raw pointers). Unlike the other two types, it is not suitable for long-term storage of peers.

The three types address three different needs in the current official code. This PR introduces two of them. Another PR (already in the works) will introduce CachePeers. Eventually, these types will change and some may disappear, but different kinds of CachePeer collections will remain for the foreseeable future.

> "selected" has no relevance at the level of type definition - it should be in the local variable name for callers where that property is true/relevant.

Selection will always be true/relevant for this type. I am open to spelling improvements, of course, but CachePeers is not a good spelling option for SelectedCachePeers because we want to emphasize that these are not necessarily all configured peers and because CachePeers spelling is best used for the primary collection of all configured CachePeer objects (as detailed in the first bullet above). And the difference between the two types is not just about a subset of peers or CachePeer ownership. There is a lot more baggage that the type in the first bullet has to deal with and other types do not care about.